In [14]:
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Flatten, Dense, TimeDistributed

### Import The Data

In [15]:
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words = 10000)

<string>:6: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/imdb.py:159: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/datasets/imdb.py:160: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 

### Padding Each Scentence to the same length

In [16]:
# Maxiumum sequence length 300

X_train = pad_sequences(X_train, maxlen = 300)
X_test =  pad_sequences(X_test, maxlen = 300)

### Shape of Features and Labels

#### Features

In [17]:
print("Number of review, number of words in each review in the Training Data")
print(X_train.shape)
print("Number of review, number of words in each review in the Testing Data")
print(X_test.shape)

Number of review, number of words in each review in the Training Data
(25000, 300)
Number of review, number of words in each review in the Testing Data
(25000, 300)


#### Labels

In [5]:
print("Number of Labels in the Training Data")
print(y_train.shape)
print("Number of Labels in the Training Data")
print(y_test.shape)

Number of Labels in the Training Data
(25000,)
Number of Labels in the Training Data
(25000,)


### Print value of any one feature and it's label 

#### Lets Check a Feature and label of a record with its index

In [18]:
X_train[30]

array([   0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    1,   13, 1059,   14,   33,    6,
       2181, 2824,   32,   13,   70,  135,    9,   51,    4,  609,   14,
         20,  299,   46,   44,   17, 9287,   17, 41

In [19]:
y_train[30]

0

- We can see the label of the entry 30 is 0 which is negative

In [20]:
X_train[652]

array([  93,    4, 1985, 6751,    5,   11,  359,    7,  766,   38,   29,
       8929,    4,   86,  255,   29,   70,  103, 6518,    4,   85,  605,
          9,    6,  601,  136,    7,    2,   11, 1445,   19, 4115,    5,
         59,    9,  110, 1298,    6,   55,  729, 4094,  136,   16,  605,
         46,    7,    4,  178,  310,    7, 1985, 2694, 1985, 1918,    4,
       1362,  255,   17,   73,    4,   22,  127,   28,   45,  712,  151,
         18,  813,    4,  167,   16, 1819,    4,   78,  759,   11,   18,
          4, 1985,   33,  757,    4, 2526,  631, 3110,    4, 7166,    7,
        844,   39,    4,   86,    2,   22,  951,    7,    4,    2, 2544,
          2, 1775, 2573,    5,    4, 5367,    2,    7,   15,  598,  228,
        603,  469,  525,   21,  110,   11,   45,  204, 5633, 2718,    5,
       5444, 5987, 1985,  115, 6026,   12,    9,   31,    7,    4,    2,
          5,   91, 3599,    7,    4,    2, 1985,  201,   19,    6,  114,
        347,    2,   11,   45,  285,   21,    4, 39

In [21]:
y_train[652]

1

- We can notice the label for the entry at 652 is 1 which is positive

### Creating Word index

In [22]:
word_index = imdb.get_word_index()

1646592/1641221 [==============================] - 0s 0us/step


#### lets ge the original value of the encoding from the word index dictionary

In [24]:
for encoding in X_train[30]:
  for key, value in word_index.items():
    if encoding == value:
      print(key, end = " ")

the was pace as they is underrated expression an was well why it when of number as on effects some has movie darling movie discuss fallen leaves of guy but nbc if super predictable minutes of territory far br actor to which what have people cannot man literally this of late of you it these her super for that interesting which makes i i of little crowd acid fame ex bryan looks in looked is acid really places is notably and of notably trying watch bryan in also were and those bryan whoever any but and to oscar some in known any i i series has an of their more it of hollywood it show of try and br structure starring and can and to and to nowadays waters without move they by come bryan and is and and to plague accurate film is notably park are try br that title series times ex bryan plague accurate is feel are try br so and episode for at man reynolds but as hugh do you 

In [25]:
y_train[30]

0

In [26]:
for encoding in X_train[652]:
  for key, value in word_index.items():
    if encoding == value:
      print(key, end = " ")

way of regular pumbaa to this kids br storyline her all examined of how found all well watch pale of because ok it is group scenes br and this happening film financial to would it life touching is time sets passes scenes with ok some br of want house br regular con regular manage of ride found movie much of you end one if words old but soundtrack of going with originally of do bunch this but of regular they mention of walked hours per of moss br move or of how and you cheesy br of and magical and exist subtitles to of roof and br for complete making simple unfortunately won't not life this if i've studies initially to shoulder elevator regular best gillian that it by br of and to its ticket br of and regular original film is little top and this if dvd not of sadness garbo fellow of ok and blond to and truly goes taxi br as you people that even and to – series of house family seeing colin are of you one life of you remote is affection and to final wood of li storyline terrific are of sh

In [27]:
y_train[652]

1

- we can notice that for the record in index 30 the label is 0 which is negative and we can clearly see the original encoding
- For the record 652 the label is positive(1) which is shown above

### Model Building

In [28]:
# Sequential model
model = Sequential()

# Embedding Layer - Keras , Vocabulary size = 10000, dimension of the dense embedding = 100, length of the input sequence  = 300
model.add(Embedding(input_dim = 10000, output_dim = 100, input_length = 300))

# LSTM with return seqence True with activation function relu
model.add(LSTM(64, activation = 'relu', return_sequences=True))

# TimeDistributed with 100 dense neurons
model.add(TimeDistributed(Dense(100)))

# Flatten Layer
model.add(Flatten())

# Dense Layer
model.add(Dense(1024, activation = 'relu'))
model.add(Dense(512, activation = 'relu'))
model.add(Dense(256, activation = 'relu'))


# last dense layer used sigmoid funtion for giving output o or 1
model.add(Dense(1, activation='sigmoid'))

### Compile The model

In [29]:
# optimier used adam, loss - binary cross etropy, accuracy is used as metrics
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

### Summary of the model

In [ ]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding (Embedding)        (None, 300, 100)          1000000   
_________________________________________________________________
lstm (LSTM)                  (None, 300, 64)           42240     
_________________________________________________________________
time_distributed (TimeDistri (None, 300, 100)          6500      
_________________________________________________________________
flatten (Flatten)            (None, 30000)             0         
_________________________________________________________________
dense_1 (Dense)              (None, 1024)              30721024  
_________________________________________________________________
dense_2 (Dense)              (None, 512)               524800    
_________________________________________________________________
dense_3 (Dense)              (None, 256)               1

- relu is uded on the hidden layers
- the units are fully connected because of dense layer
- we use sigmoid on output layer which gives either 0 or 1

In [32]:
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=300)

Epoch 1/10
84/84 [==============================] - 256s 3s/step - loss: 0.6176 - accuracy: 0.6059 - val_loss: 0.3336 - val_accuracy: 0.8662
Epoch 2/10
84/84 [==============================] - 254s 3s/step - loss: 0.2006 - accuracy: 0.9248 - val_loss: 0.3020 - val_accuracy: 0.8805
Epoch 3/10
84/84 [==============================] - 255s 3s/step - loss: 0.1088 - accuracy: 0.9626 - val_loss: 0.3719 - val_accuracy: 0.8684
Epoch 4/10
84/84 [==============================] - 253s 3s/step - loss: 0.0462 - accuracy: 0.9859 - val_loss: 0.4969 - val_accuracy: 0.8648
Epoch 5/10
84/84 [==============================] - 256s 3s/step - loss: 0.0215 - accuracy: 0.9930 - val_loss: 0.6665 - val_accuracy: 0.8554
Epoch 6/10
84/84 [==============================] - 256s 3s/step - loss: 0.0155 - accuracy: 0.9950 - val_loss: 0.8542 - val_accuracy: 0.8621
Epoch 7/10
84/84 [==============================] - 255s 3s/step - loss: 0.0087 - accuracy: 0.9972 - val_loss: 0.9353 - val_accuracy: 0.8571
Epoch 8/10
84

- The model overfits on the third epoch and the validation accuracy drops after that.
- The model will be good if we stop on epoch 2

#### Evaluate The Above Model

In [37]:
model_score = model.evaluate(X_test, y_test)

782/782 [==============================] - 74s 95ms/step - loss: 1.1915 - accuracy: 0.8604


In [38]:
print("The loss of the model is : " , model_score[0])
print("The Accuracy of the model is : ", model_score[1])

The loss of the model is :  1.191495418548584
The Accuracy of the model is :  0.8604000210762024


- loss is very high

### Redefining The model again with same structure

In [39]:
# Sequential model
model1 = Sequential()

# Embedding Layer - Keras , Vocabulary size = 10000, dimension of the dense embedding = 100, length of the input sequence  = 300
model1.add(Embedding(input_dim = 10000, output_dim = 100, input_length = 300))

# LSTM with return seqence True with activation function relu
model1.add(LSTM(64, activation = 'relu', return_sequences=True))

# TimeDistributed with 100 dense neurons
model1.add(TimeDistributed(Dense(100)))

# Flatten Layer
model1.add(Flatten())

# Dense Layer
model1.add(Dense(1024, activation = 'relu'))
model1.add(Dense(512, activation = 'relu'))
model1.add(Dense(256, activation = 'relu'))
model1.add(Dense(1, activation='sigmoid'))

#### Compile Model1

In [44]:
# optimier used adam, loss - binary cross etropy, accuracy is used as metrics
model1.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

### Summary of Model1

In [45]:
model1.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_2 (Embedding)      (None, 300, 100)          1000000   
_________________________________________________________________
lstm_2 (LSTM)                (None, 300, 64)           42240     
_________________________________________________________________
time_distributed_2 (TimeDist (None, 300, 100)          6500      
_________________________________________________________________
flatten_2 (Flatten)          (None, 30000)             0         
_________________________________________________________________
dense_11 (Dense)             (None, 1024)              30721024  
_________________________________________________________________
dense_12 (Dense)             (None, 512)               524800    
_________________________________________________________________
dense_13 (Dense)             (None, 256)              

In [46]:
model1.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=2, batch_size=300)

Epoch 1/2
84/84 [==============================] - 256s 3s/step - loss: 0.5802 - accuracy: 0.6369 - val_loss: 0.2843 - val_accuracy: 0.8790
Epoch 2/2
84/84 [==============================] - 255s 3s/step - loss: 0.1969 - accuracy: 0.9267 - val_loss: 0.2960 - val_accuracy: 0.8767


- The above model is not overfitting hence 2 epochs is very good

#### Evaluation of Model1

In [47]:
model1_score = model1.evaluate(X_test, y_test)

782/782 [==============================] - 74s 95ms/step - loss: 0.2960 - accuracy: 0.8767


In [48]:
print("The loss of the model1 is : " , model1_score[0])
print("The Accuracy of the model1 is : ", model1_score[1])

The loss of the model1 is :  0.2959861755371094
The Accuracy of the model1 is :  0.8766800165176392


### Predict using model

In [52]:
model1.predict(X_test[30].reshape((1, 300)))

array([[0.9582788]], dtype=float32)

In [54]:
y_test[30]

1

In [53]:
model.predict(X_test[652].reshape((1, 300)))

array([[3.7246295e-09]], dtype=float32)

In [55]:
y_test[652]

0

- The model is predicting well. 
- The above 2 samples of record 30 and 652 is predicted well in the test set and are accurate
- our model has validation accuracy of 87% and train accuracy of 92% 